File: packages/goalos-py/goalos/permissions.py
"""
Permission management for agent access control
"""

from typing import List, Optional

from .types import (
    Goal,
    Permission,
    PermissionCapability,
    PermissionScope,
)


class PermissionManager:
    """Manages permissions for agent access to goals"""

    @staticmethod
    def grant_permission(
        agent_id: str,
        capabilities: List[PermissionCapability],
        scope: Optional[PermissionScope] = None,
        permissions: Optional[List[Permission]] = None,
    ) -> Permission:
        """Create a new permission for an agent"""
        return Permission(
            agentId=agent_id,
            capabilities=capabilities,
            scope=scope,
        )

    @staticmethod
    def revoke_permission(
        agent_id: str,
        permissions: Optional[List[Permission]] = None,
    ) -> List[Permission]:
        """Revoke all permissions for an agent"""
        if not permissions:
            return []
        return [p for p in permissions if p.agentId != agent_id]

    @staticmethod
    def revoke_all_permissions(
        agent_id: str,
        permissions: Optional[List[Permission]] = None,
    ) -> List[Permission]:
        """Revoke all permissions for an agent"""
        return PermissionManager.revoke_permission(agent_id, permissions)

    @staticmethod
    def get_agent_permissions(
        agent_id: str,
        permissions: Optional[List[Permission]] = None,
    ) -> Optional[Permission]:
        """Get permissions for a specific agent"""
        if not permissions:
            return None
        return next(
            (p for p in permissions if p.agentId == agent_id),
            None
        )

    @staticmethod
    def has_capability(
        agent_id: str,
        capability: PermissionCapability,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if an agent has a specific capability for a goal"""
        if not permissions:
            return False

        # Find agent permissions from both goal and default
        agent_perms = [p for p in permissions if p.agentId == agent_id]
        if not agent_perms:
            return False

        # Check if any permission grants the capability
        for perm in agent_perms:
            if capability not in perm.capabilities:
                continue

            # Check scope constraints
            if perm.scope:
                if perm.scope.goalIds and goal.id not in perm.scope.goalIds:
                    continue
                if perm.scope.domains and goal.domain and goal.domain not in perm.scope.domains:
                    continue
                # maxDepth would need parent chain calculation

            return True

        return False

    @staticmethod
    def can_read(
        agent_id: str,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if agent can read a goal"""
        return PermissionManager.has_capability(agent_id, "read", goal, permissions)

    @staticmethod
    def can_write(
        agent_id: str,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if agent can write to a goal"""
        return PermissionManager.has_capability(agent_id, "write", goal, permissions)

    @staticmethod
    def can_complete(
        agent_id: str,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if agent can complete a goal"""
        return PermissionManager.has_capability(agent_id, "complete", goal, permissions)

    @staticmethod
    def can_create_sub_goals(
        agent_id: str,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if agent can create sub-goals"""
        return PermissionManager.has_capability(agent_id, "create_sub_goals", goal, permissions)

    @staticmethod
    def can_reprioritize(
        agent_id: str,
        goal: Goal,
        permissions: Optional[List[Permission]] = None,
    ) -> bool:
        """Check if agent can reprioritize a goal"""
        return PermissionManager.has_capability(agent_id, "reprioritize", goal, permissions)

    @staticmethod
    def filter_goals_readable(
        agent_id: str,
        goals: List[Goal],
        permissions: Optional[List[Permission]] = None,
    ) -> List[Goal]:
        """Filter goals to only those readable by agent"""
        return [
            g for g in goals
            if PermissionManager.can_read(agent_id, g, permissions)
        ]

    @staticmethod
    def filter_goals_writable(
        agent_id: str,
        goals: List[Goal],
        permissions: Optional[List[Permission]] = None,
    ) -> List[Goal]:
        """Filter goals to only those writable by agent"""
        return [
            g for g in goals
            if PermissionManager.can_write(agent_id, g, permissions)
        ]

File: packages/goalos-core/src/permissions.ts
/**
 * Permission scoping and authorization for agents
 */

import type { Permission, PermissionCapability, PermissionScope, Goal } from './types.js';
import { buildGoalMap, getAncestorIds, getDescendantIds } from './utils.js';

/**
 * Permission manager for controlling agent access
 */
export class PermissionManager {
  /**
   * Check if an agent has a specific capability on a goal
   */
  static hasCapability(
    agentId: string,
    capability: PermissionCapability,
    goal: Goal,
    permissions: Permission[]
  ): boolean {
    const agentPermission = permissions.find(p => p.agentId === agentId);
    if (!agentPermission) {
      return false;
    }

    if (!agentPermission.capabilities.includes(capability)) {
      return false;
    }

    // Check scope constraints
    if (agentPermission.scope) {
      return this.checkScope(goal.id, agentPermission.scope, goal, permissions);
    }

    return true;
  }

  /**
   * Check if a goal is within the permission scope
   */
  static checkScope(
    goalId: string,
    scope: PermissionScope,
    goal: Goal,
    permissions: Permission[]
  ): boolean {
    if (scope.goalIds && !scope.goalIds.includes(goalId)) {
      return false;
    }

    if (scope.domains && goal.domain && !scope.domains.includes(goal.domain)) {
      return false;
    }

    return true;
  }

  /**
   * Grant a permission to an agent
   */
  static grantPermission(
    agentId: string,
    capabilities: PermissionCapability[],
    scope?: PermissionScope,
    existing?: Permission[]
  ): Permission {
    return {
      agentId,
      capabilities,
      scope
    };
  }

  /**
   * Revoke all permissions for an agent
   */
  static revokeAllPermissions(agentId: string, permissions: Permission[]): Permission[] {
    return permissions.filter(p => p.agentId !== agentId);
  }

  /**
   * Revoke a specific capability
   */
  static revokeCapability(
    agentId: string,
    capability: PermissionCapability,
    permissions: Permission[]
  ): Permission[] {
    return permissions.map(p => {
      if (p.agentId === agentId) {
        return {
          ...p,
          capabilities: p.capabilities.filter(c => c !== capability)
        };
      }
      return p;
    });
  }

  /**
   * Add capabilities to existing permission
   */
  static addCapabilities(
    agentId: string,
    capabilities: PermissionCapability[],
    permissions: Permission[]
  ): Permission[] {
    return permissions.map(p => {
      if (p.agentId === agentId) {
        const newCapabilities = [...new Set([...p.capabilities, ...capabilities])];
        return {
          ...p,
          capabilities: newCapabilities
        };
      }
      return p;
    });
  }

  /**
   * Get all goals an agent can read
   */
  static getReadableGoals(agentId: string, goals: Goal[], globalPermissions: Permission[]): Goal[] {
    return goals.filter(goal => this.hasCapability(agentId, 'read', goal, globalPermissions));
  }

  /**
   * Get all goals an agent can modify
   */
  static getWritableGoals(agentId: string, goals: Goal[], globalPermissions: Permission[]): Goal[] {
    return goals.filter(goal => this.hasCapability(agentId, 'write', goal, globalPermissions));
  }

  /**
   * Get all goals an agent can complete
   */
  static getCompletableGoals(agentId: string, goals: Goal[], globalPermissions: Permission[]): Goal[] {
    return goals.filter(goal => this.hasCapability(agentId, 'complete', goal, globalPermissions));
  }

  /**
   * Check if agent can create sub-goals
   */
  static canCreateSubGoals(agentId: string, goal: Goal, globalPermissions: Permission[]): boolean {
    return this.hasCapability(agentId, 'create_sub_goals', goal, globalPermissions);
  }

  /**
   * Check if agent can reprioritize
   */
  static canReprioritize(agentId: string, goal: Goal, globalPermissions: Permission[]): boolean {
    return this.hasCapability(agentId, 'reprioritize', goal, globalPermissions);
  }

  /**
   * Validate permission structure
   */
  static validatePermission(permission: Permission): { valid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!permission.agentId || permission.agentId.trim() === '') {
      errors.push('Agent ID is required');
    }

    if (!permission.capabilities || permission.capabilities.length === 0) {
      errors.push('At least one capability is required');
    }

    const validCapabilities: PermissionCapability[] = ['read', 'write', 'complete', 'create_sub_goals', 'reprioritize'];
    for (const cap of permission.capabilities) {
      if (!validCapabilities.includes(cap)) {
        errors.push(`Invalid capability: ${cap}`);
      }
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }

  /**
   * Create a permission with depth-limited scope
   */
  static createDepthLimitedPermission(
    agentId: string,
    capabilities: PermissionCapability[],
    maxDepth: number
  ): Permission {
    return {
      agentId,
      capabilities,
      scope: { maxDepth }
    };
  }

  /**
   * Create a domain-scoped permission
   */
  static createDomainScopedPermission(
    agentId: string,
    capabilities: PermissionCapability[],
    domains: string[]
  ): Permission {
    return {
      agentId,
      capabilities,
      scope: { domains }
    };
  }

  /**
   * Create a goal-specific permission
   */
  static createGoalScopedPermission(
    agentId: string,
    capabilities: PermissionCapability[],
    goalIds: string[]
  ): Permission {
    return {
      agentId,
      capabilities,
      scope: { goalIds }
    };
  }

  /**
   * Get all agents with a specific capability
   */
  static getAgentsWithCapability(
    capability: PermissionCapability,
    permissions: Permission[]
  ): string[] {
    return permissions
      .filter(p => p.capabilities.includes(capability))
      .map(p => p.agentId);
  }

  /**
   * Get all permissions for an agent
   */
  static getAgentPermissions(agentId: string, permissions: Permission[]): Permission | undefined {
    return permissions.find(p => p.agentId === agentId);
  }

  /**
   * Check if any agent has write access
   */
  static hasAnyWriter(permissions: Permission[]): boolean {
    return permissions.some(p => p.capabilities.includes('write'));
  }

  /**
   * Get summary of who can do what
   */
  static getSummary(permissions: Permission[]): Record<string, PermissionCapability[]> {
    const summary: Record<string, PermissionCapability[]> = {};
    for (const permission of permissions) {
      summary[permission.agentId] = permission.capabilities;
    }
    return summary;
  }
}
